ee/clickhouse/queries/clickhouse_retention.py

```diff
@@ -58,62 +50,103 @@ def calculate_retention(self, filter: Filter, team: Team, total_intervals: int)
+                team_id=team.pk,
+            )
+            .c(FROM=f"events e JOIN ({person_q}) pdi on e.distinct_id = pdi.distinct_id")
+            .update_params(**person_q.params),
+            lambda q: target_query and q.c(WHERE=target_query).update_params(**target_params),
+            lambda q: prop_filters and q.c(WHERE=prop_filters).update_params(**prop_filter_params),
+        )
+
+        reference_event_query = compose(
+            lambda: SqlQuery(
+                SELECT="""
+                        DISTINCT pdi.person_id as person_id,
+                        {trunc_func}(e.timestamp) as event_date
+                        """.format(
+                    trunc_func=trunc_func
+                )
+            )
+            .c(
+                WHERE="""
+                        toDateTime(e.timestamp) >= toDateTime(%(start_date)s)
+                        AND toDateTime(e.timestamp) <= toDateTime(%(end_date)s)
+                        AND e.team_id = %(team_id)s
+                        """
+            )
+            .update_params(
+                start_date=date_from.strftime(
                     "%Y-%m-%d{}".format(" %H:%M:%S" if filter.period == "Hour" else " 00:00:00")
                 ),
-                **prop_filter_params,
-                **target_params,
-                "period": period,
-            },
+                end_date=date_to.strftime("%Y-%m-%d{}".format(" %H:%M:%S" if filter.period == "Hour" else " 00:00:00")),
+                team_id=team.pk,
+            )
+            .c(FROM=f"events e JOIN ({person_q}) pdi on e.distinct_id = pdi.distinct_id")
+            .update_params(**person_q.params),
+            lambda q: target_query and q.c(WHERE=target_query).update_params(**target_params),
+            lambda q: prop_filters and q.c(WHERE=prop_filters).update_params(**prop_filter_params),
         )
 
+        final_query = SqlQuery(
+            SELECT="""
+                datediff(%(period)s, {trunc_func}(toDateTime(%(start_date)s)), reference_event.event_date) as period_to_event_days,
+                datediff(%(period)s, reference_event.event_date, {trunc_func}(toDateTime(event_date))) as period_between_events_days,
+                COUNT(DISTINCT event.person_id) count
+            """.format(
+                trunc_func=trunc_func
+            ),
+            FROM=f"({event_query}) event JOIN ({reference_event_query}) reference_event ON (event.person_id = reference_event.person_id)",
+            WHERE="{trunc_func}(event.event_date) >= {trunc_func}(reference_event.event_date)".format(
+                trunc_func=trunc_func
+            ),
+            GROUP="period_to_event_days, period_between_events_days",
+            ORDER="period_to_event_days, period_between_events_days",
+        ).update_params(period=period, **reference_event_query.params)
+
+        result = sync_execute(str(final_query), final_query.params)
```

Comment:
@macobo take a look at this utilization of the query builder. so I think it makes the query cleaner, however, because these queries are really large when they get composed like this, it's much harder to see the overall shape

<img width="1148" alt="Screen Shot 2020-11-06 at 1 04 28 PM" src="https://user-images.githubusercontent.com/13127476/98364244-7204bf00-1ffe-11eb-85a6-f1ba15b81dcd.png">

<img width="979" alt="Screen Shot 2020-11-06 at 1 04 16 PM" src="https://user-images.githubusercontent.com/13127476/98364240-6f09ce80-1ffe-11eb-98ca-7442af630bc0.png">

Reply:
Good point.

I lean towards many short functions > large functions, I think we can do this here.

My suggestion with this is extracting parts of the code (similar to what Events model things did previously) into small independent functions that live outside of this one large function.

For example:

```python
def persons_query(team_id):
    return SqlQuery(
            {	            SELECT="""
                "team_id": team.pk,	                person_id,
                "start_date": date_from.strftime(	                distinct_id
            """,
            FROM="""
                person_distinct_id
            """,
            WHERE="""
                team_id = %(team_id)s
            """,
        ).update_params(team_id=team.id)

def events_in_time_range_query(...)

def reference_event_query(...)
```

And in the main function we only compose the parts as you do above, calling into the subqueries.

Lastly, to see the actual query created, I think snapshot tests will come in handy there - we can drop in the generated snapshot file and see what the actual query looks like.